conn.go

```diff
@@ -474,8 +483,12 @@ func (s *startupCoordinator) startup(ctx context.Context, supported map[string][
 	case error:
 		return v
 	case *readyFrame:
+		// Connection is successfully set up and ready to use Native Protocol v5
+		s.conn.startupCompleted = true
```

Comment:
Don't we need some kind of sychronization around this field? Or atomic access?

Reply:
As I see `Conn.startupCompleted` is being read by:

- [Conn.serve](https://github.com/apache/cassandra-gocql-driver/blob/953e0df999cabb3f5eef714df9921c00e9f632c2/conn.go#L587) which is run by [Conn.init](https://github.com/apache/cassandra-gocql-driver/blob/953e0df999cabb3f5eef714df9921c00e9f632c2/conn.go#L306) after `startup.setupConn` is finished. [startupCoordinator.setupConn](https://github.com/apache/cassandra-gocql-driver/blob/953e0df999cabb3f5eef714df9921c00e9f632c2/conn.go#L372) sends **STARTUP** as not framed envolope, receives **READY** or **AUTHENTICATE** and sets `Conn.startupCompleted` to true.
`Conn.serve` is running in a separate go routine
- [Conn.exec](https://github.com/apache/cassandra-gocql-driver/blob/953e0df999cabb3f5eef714df9921c00e9f632c2/conn.go#L1029) which sends queries to C* and returns `framer` which contains C* response.
`Conn.exec` is also running in different go routines.

As I see there are no sync issues because `Conn.serve` is run after the `startupCoordinator.startup` is completed, so it already reads messages from C* by using proto v5. Even if during startup it receives **AUTHENTICATE** response from C*, it will respond with **AUTH_RESPONSE** using proto v5.

Reply:
Yes but if `startupCompleted` is not `atomic` then there is no guarantee that the goroutine that reads this field will get the most recent value since the goroutine that writes it is a different one no?

Reply:
`startupCoordinator.setupConn` is complex, so I'll add its code to this message for clarification
```go
func (s *startupCoordinator) setupConn(ctx context.Context) error {
	var cancel context.CancelFunc
	if s.conn.timeout > 0 {
		ctx, cancel = context.WithTimeout(ctx, s.conn.timeout)
	} else {
		ctx, cancel = context.WithCancel(ctx)
	}
	defer cancel()

	startupErr := make(chan error)
	go func() { // the reading goroutine
		for range s.frameTicker {
			err := s.conn.recv(ctx)
			if err != nil {
				select {
				case startupErr <- err:
				case <-ctx.Done():
				}

				return
			}
		}
	}()

	go func() { // the writing one
		defer close(s.frameTicker)
		err := s.options(ctx)
		select {
		case startupErr <- err:
		case <-ctx.Done():
		}
	}()

	select {
	case err := <-startupErr:
		if err != nil {
			return err
		}
	case <-ctx.Done():
		return errors.New("gocql: no response to connection startup within timeout")
	}

	return nil
}
```

It runs two goroutines: the first listening for response from C* and the second for processing startup.
The first gorounite starts listening data from C* when the `startupCoordinator.frameTicker` chan is not empty.
The second goroutine sends queries to C* via [startupCoordinator.write](https://github.com/apache/cassandra-gocql-driver/blob/953e0df999cabb3f5eef714df9921c00e9f632c2/conn.go#L417) which also writes to `startupCoordinator.frameTicker` chan. So, the reading from C* only starts when the **writing gouroutine** sends a query to the C*, the both goroutines are synchronized by `startupCoordinator.frameTicker`. Once the startup is completed, those goroutines are finished.

When `startupCoordinator.setupConn` is finished, `Conn.init` runs two new goroutines: one for connection serving, and another one for a heartbeat`, so they both already have an updated state of `startupCompleted`.

Reply:
I think there's a way to avoid using `atomic`... You'd have to remove the field from the `Conn` object completely and pass it to `exec()` and `recv()` as an argument which doesn't seem too complicated to do. Maybe `exec` can become a wrapper around an `internalExec` so we don't have to set `startupCompleted` on all places that call `exec` where `startupCompleted` is always `true` (like `executeQuery`, `executeBatch`, `heartBeat`, etc.)

Reply:
Ok I understand, the `startupErr` and `frameTicker` channels ensure that everything is properly synchronized before we return the `conn` object so there's no race.

I still think this is very obscure as it stands, having shared state like this is not great... Is there any other field that is being "implictly" synchronized like this? `frameTicker` wasn't really meant to synchronize shared state if I understand correctly, it was probably just meant to notify `recv()` that a response is expected.

I can easily see a bug being introduced in the future because the visibility of `startupCompleted` is being ensured by channel operations that have a different main purpose and I doubt a dev looking at this code for the first time will recognize this.

Reply:
> Is there any other field that is being "implictly" synchronized like this?

Specifically, I did this way only for `startupCompleted` and any other occurrences I didn't meet

> I doubt a dev looking at this code for the first time will recognize this.

Oh, I didn't look from this side... Then we could use atomics here, but the whole implementation won't change much. It still will update the state of `startupCompleted` inside during `startup()` for the **AUTHENTICATE** message if there it is. Or maybe implicitly call `recv()` the second time... Ok, I'll look at what I can do here.

Reply:
How do you feel about [this](https://github.com/joao-r-reis/gocql/commit/6b6ecf5b857da57d1068a3af547b35430ec7b496), I must admit I didn't test this at all but if this works then we don't have a shared state without explicit synchronization AND we don't have to perform atomic operations after the handshake is done so no performance penalty.

Reply:
It looks good to me. It provides atomic access and notifies if the startup is completed for auth message handling as well. I could take it and test

Reply:
I tested this approach and everything is working fine. It passed integration and unit tests successfully.